diff --git a/hyprscrolling/Scrolling.cpp b/hyprscrolling/Scrolling.cpp
index 39e6717..f2215af 100644
--- a/hyprscrolling/Scrolling.cpp
+++ b/hyprscrolling/Scrolling.cpp
@@ -1468,3 +1468,15 @@ SP<SWorkspaceData> CScrollingLayout::currentWorkspaceData() {
 CBox CScrollingLayout::usableAreaFor(PHLMONITOR m) {
     return CBox{m->m_reservedTopLeft, m->m_size - m->m_reservedTopLeft - m->m_reservedBottomRight};
 }
+
+void CScrollingLayout::onWindowFocusChange(PHLWINDOW wh) {
+    if (!wh)
+        return;
+    const auto DATA = dataFor(wh);
+    const auto WS   = dataFor(wh->m_workspace);
+    const auto COL  = DATA->column.lock();
+    if (!DATA || !WS || !COL)
+        return;
+    WS->centerOrFitCol(COL);
+    WS->recalculate();
+}
diff --git a/hyprscrolling/Scrolling.hpp b/hyprscrolling/Scrolling.hpp
index 21b5dbc..e45c5a9 100644
--- a/hyprscrolling/Scrolling.hpp
+++ b/hyprscrolling/Scrolling.hpp
@@ -105,6 +105,8 @@ class CScrollingLayout : public IHyprLayout {
 
     CBox                             usableAreaFor(PHLMONITOR m);
 
+    void                             onWindowFocusChange(PHLWINDOW wh);
+
   private:
     std::vector<SP<SWorkspaceData>> m_workspaceDatas;
 
